### PR TITLE
Trigger container cleanup within a pod when a container exiting event is detected

### PR DESCRIPTION
**Pull request by ronnielai**
#25239

### CODE REVIEW

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -152,6 +152,8 @@ const (
 	ContainerGCPeriod = time.Minute
 	// Period for performing image garbage collection.
 	ImageGCPeriod = 5 * time.Minute
+	// Max number of dead containers to keep in a pod
+	maxDeadContainerInPod = 1
```
Comment (vishh) [inline pkg/kubelet/kubelet.go]: Is this max or min?

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -478,6 +480,7 @@ func NewMainKubelet(
 		return nil, err
 	}
 	klet.containerGC = containerGC
+	klet.containerDeletionWorker = newContainerDeletionWorker(klet.containerRuntime)
```
Comment (vishh) [inline pkg/kubelet/kubelet.go]: Ideally, the deletion worker should be part of the docker runtime implementation right?

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -3567,7 +3579,18 @@ func (kl *Kubelet) ListenAndServeReadOnly(address net.IP, port uint) {
 	server.ListenAndServeKubeletReadOnlyServer(kl, kl.resourceAnalyzer, address, port, kl.containerRuntime)
 }
 
-// Filter out events that are not worthy of pod syncing
+// deleteContainersInPod issues container deletion requests
+func (kl *Kubelet) deleteContainersInPod(candidates ContainerStatusbyCreated) {
+	for _, candidate := range candidates {
+		select {
+		case kl.containerDeletionWorker <- candidate.ID:
```
Comment (vishh) [inline pkg/kubelet/kubelet.go]: Is this expected to be blocking? What is the interface here? Fire and continue or wait for success?

File: pkg/kubelet/util.go [hunk 4]
```diff
@@ -118,3 +127,56 @@ func allowHostIPC(pod *api.Pod) (bool, error) {
 	}
 	return false, nil
 }
+
+type ContainerStatusbyCreated []*kubecontainer.ContainerStatus
```
Comment (vishh) [inline pkg/kubelet/util.go]: nit: include `list` in the type name to make it readable.

File: pkg/kubelet/util.go [hunk 5]
```diff
@@ -118,3 +127,56 @@ func allowHostIPC(pod *api.Pod) (bool, error) {
...
+			continue
+		}
+		if containerStatus.ID.ID == exitedContainerID {
+			matchedContainer = containerStatus
+		}
+		exitedContainers = append(exitedContainers, containerStatus)
+	}
+	if matchedContainer == nil {
+		glog.Warningf("Container %q not found in pod's exited containers", exitedContainerID)
+		return ContainerStatusbyCreated{}
```
Comment (vishh) [inline pkg/kubelet/util.go]: Why are we returning here even if there are more containers than can be deleted?

File: pkg/kubelet/util.go [hunk 6]
```diff
@@ -18,11 +18,20 @@ package kubelet
...
 	"k8s.io/kubernetes/pkg/capabilities"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
 	kubetypes "k8s.io/kubernetes/pkg/kubelet/types"
 	"k8s.io/kubernetes/pkg/securitycontext"
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+const (
+	// The limit on the number of buffered container deletion requests
+	containerDeletionWorkerBufferLimit = 10
```
Comment (vishh) [inline pkg/kubelet/util.go]: How did you arrive at this limit?

File: pkg/kubelet/kubelet.go [hunk 7]
```diff
@@ -3567,7 +3579,18 @@ func (kl *Kubelet) ListenAndServeReadOnly(address net.IP, port uint) {
...
+		default:
+			glog.Warningf("Failed to issue the reqeust to remove container %v", candidate.ID)
```
Comment (vishh) [inline pkg/kubelet/kubelet.go]: nit: s/reqeust/request

**Comment by vishh** [comment]
Added a few comments. Overall LGTM.

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (ronnielai) [inline pkg/kubelet/kubelet.go]: It's the max. It represents the maximum dead containers that'll be kept.

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (ronnielai) [inline pkg/kubelet/kubelet.go]: The runtime implementation provides a way to delete a specific container through DeleteContainer. The worker here is kind of an optimization. I use the worker to represent a go routine that blocks on DeleteContainer calls. Another simpler approach is to scratch the worker idea and just do "go DeleteContainer(...)" in line 2358. Please advise.

File: pkg/kubelet/kubelet.go (hunk 3 above)
Comment (ronnielai) [inline pkg/kubelet/kubelet.go]: It's non-blocking (fire and continue) to protect the syncloop. If the channel is full, the default case is triggered.

File: pkg/kubelet/util.go (hunk 4 above)
Comment (ronnielai) [inline pkg/kubelet/util.go]: Will do

File: pkg/kubelet/util.go (hunk 5 above)
Comment (ronnielai) [inline pkg/kubelet/util.go]: We need to get the name of the exited container to find all the exited containers with the same name. This is a case that should never happen. (If it happens, it means pod cache is not updated properly.)

File: pkg/kubelet/kubelet.go (hunk 7 above)
Comment (ronnielai) [inline pkg/kubelet/kubelet.go]: Will do

File: pkg/kubelet/util.go (hunk 6 above)
Comment (ronnielai) [inline pkg/kubelet/util.go]: I copied from podKillingChannelCapacity. I've added a comment to indicate the number is arbitrary.

**Comment by ronnielai** [comment]
PTAL

**Comment by ronnielai** [comment]
@vishh PTAL

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (ronnielai) [inline pkg/kubelet/kubelet.go]: Chatted off-line. Will use it as a min.

**Comment by vishh** [comment]
LGTM

**Pull request merged**